projects/2-state-machines/chern.ipynb

```diff
@@ -0,0 +1,492 @@
+   "source": [
+    "Note that any path through a state machine ending in a terminal state is a valid string, and each step of the path is associated with a symbol (i.e., a letter). To combine symbols together into strings, you can use string concatenation:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "collapsed": false
+   },
+   "outputs": [],
+   "source": [
+    "ab = \"a\" + \"b\"\n",
+    "print(ab)\n",
+    "abc = ab + \"c\"\n",
+    "print(abc)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "You can combine lists the same way:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "collapsed": false
+   },
+   "outputs": [],
+   "source": [
+    "lst_abc = [\"a\",\"b\"] + [\"c\"]\n",
+    "print(lst_abc)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "All that's left is enumerating the valid paths of a given length and assembling those into strings. The skeleton below is a _recursive_ function that calls itself repeatedly, branching out according to each option. You can imagine that the call to `sample()` is the root of a tree of calls to `sample2`, and the paths through that tree are paths through the state machine---the path itself is \"stored\" in the third argument to `sample2`. In other words:\n",
+    "\n",
+    "* To get the possible paths of length `K` from a state machine `SM`:\n",
+    "    * Initialize `S` to `SM`'s current state.\n",
+    "    * Initialize the `result` to `[]`\n",
+    "    * For each available out-edge at `S`:\n",
+    "        * Advance `SM` along that edge\n",
+    "        * Get the possible paths of length `K-1` from `SM` and add them to the result\n",
+    "        * Put `SM` back into state `S` (so we can advance it along the next available out edge)\n",
+    "    * Yield `result`\n",
+    "\n",
+    "It's up to you to figure out how to find the possible paths of length 0. Hint: How many such paths can there be for a given call to `sample2(sm, 0, sofar)`? What aspects of the state machine's status does it depend on?"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "# Starter function that calls sample2() with an empty string as argument. This is just to \"accumulate\" paths into\n",
+    "# that argument without tracking a bunch of data externally.\n",
+    "def sample(sm, length):\n",
+    "    return sample2(sm, length, \"\")\n",
+    "\n",
+    "# sample2 samples paths of length \"length\" from the state machine sm as of its current state.\n",
+    "# sofar is the path assembled so far.\n",
+    "# sample2 returns a list of paths.\n",
+    "def sample2(sm, length, sofar):\n",
+    "    if length == 0:\n",
+    "        # TODO: What path(s) should be returned? When?\n",
+    "        if sm.is_terminal:\n",
```

Comment:
`sm.is_terminal` is a function, not a Boolean property, so you need to write `if sm.is_terminal():` with the parentheses.